engine/rpcserver.go

```diff
@@ -1213,3 +1221,322 @@ func (s *RPCServer) GetAuditEvent(ctx context.Context, r *gctrpc.GetAuditEventRe
+
+	return &gctrpc.GCTScriptGenericResponse{
+		Status: "ok",
+		Data:   gctVM.ShortName() + " (" + gctVM.ID.String() + ") executed",
+	}, nil
+}
+
+// GCTScriptStop terminate a running script
+func (s *RPCServer) GCTScriptStop(ctx context.Context, r *gctrpc.GCTScriptStopRequest) (*gctrpc.GCTScriptGenericResponse, error) {
+	if !gctscript.GCTScriptConfig.Enabled {
+		return &gctrpc.GCTScriptGenericResponse{Status: gctscript.ErrScriptingDisabled.Error()}, nil
+	}
+
+	UUID, err := uuid.FromString(r.Script.UUID)
+	if err != nil {
+		return &gctrpc.GCTScriptGenericResponse{Status: "error", Data: err.Error()}, nil
+	}
+
+	if v, f := gctscript.AllVMs[UUID]; f {
+		err = v.Shutdown()
+		status := " terminated"
+		if err != nil {
+			status = err.Error()
+		}
+		return &gctrpc.GCTScriptGenericResponse{Status: "ok", Data: v.ID.String() + status}, nil
+	}
+	return &gctrpc.GCTScriptGenericResponse{Status: "error", Data: "no running script found"}, nil
+}
+
+// GCTScriptUpload upload a new script to ScriptPath
+func (s *RPCServer) GCTScriptUpload(ctx context.Context, r *gctrpc.GCTScriptUploadRequest) (*gctrpc.GCTScriptGenericResponse, error) {
+	if !gctscript.GCTScriptConfig.Enabled {
+		return &gctrpc.GCTScriptGenericResponse{Status: gctscript.ErrScriptingDisabled.Error()}, nil
+	}
+
+	fPath := filepath.Join(gctscript.ScriptPath, r.ScriptName)
+	_, err := os.Stat(fPath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+	} else {
+		if !r.Overwrite {
+			return nil, fmt.Errorf("%s script found and overwrite set to false", r.ScriptName)
+		} else if r.Overwrite {
+			f := filepath.Join(gctscript.ScriptPath, "version_history")
+			err = os.MkdirAll(f, 0770)
+			if err != nil {
+				return nil, err
+			}
+			timeString := strconv.FormatInt(time.Now().UnixNano(), 10)
+			renamedFile := filepath.Join(f, timeString+"-"+r.ScriptName)
+			err = file.Move(fPath, renamedFile)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	newFile, err := os.Create(fPath)
+	if err != nil {
+		return nil, err
+	}
+	defer newFile.Close()
+
+	_, err = newFile.Write(r.Data)
+	if err != nil {
+		return nil, err
+	}
+	err = newFile.Close()
+	if err != nil{
+		log.Errorln(log.Global, "failed to close file handle, archive removal may fail")
+	}
+
+	if r.Archived {
+		z, errZip := zip.OpenReader(fPath)
+		if errZip != nil {
+			return nil, err
+		}
+
+		defer z.Close()
+
+		for x := range z.Reader.File {
+			var zFile io.ReadCloser
+			zFile, err = z.Reader.File[x].Open()
+			if err != nil {
+				return nil, err
+			}
+
+			zipPath := filepath.Join(gctscript.ScriptPath, z.Reader.File[x].Name) // nolint:gosec
+			// We ignore gosec linter above because the code below files the file traversal bug when extracting archives
+			if !strings.HasPrefix(zipPath, filepath.Clean(gctscript.ScriptPath)+string(os.PathSeparator)) {
+				_ = zFile.Close()
+				return nil, fmt.Errorf("%s: illegal file path", fPath)
+			}
+
+			if z.Reader.File[x].FileInfo().IsDir() {
```

Comment:
Sadly this func doesn't work on Windows as z.Reader.File[x].Name is in the format of "path/file.gct" so it still considers it a file